Content.Client/Orbit/OrbitVisualsSystem.cs

```diff
@@ -0,0 +1,146 @@
+﻿using Content.Shared.Follower;
+using Content.Shared.Follower.Components;
+using Robust.Client.Animations;
+using Robust.Client.GameObjects;
+using Robust.Shared.Animations;
+using Robust.Shared.Random;
+
+namespace Content.Client.Orbit;
+
+public sealed class OrbitVisualsSystem : VisualizerSystem<OrbitVisualsComponent>
+{
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+
+    private readonly string _orbitAnimationKey = "orbiting";
+    private readonly string _orbitStopKey = "orbiting_stop";
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<OrbitVisualsComponent, ComponentInit>(OnComponentInit);
+        SubscribeLocalEvent<OrbitVisualsComponent, AnimationCompletedEvent>(OnAnimationCompleted);
+    }
+
+    private void OnComponentInit(EntityUid uid, OrbitVisualsComponent component, ComponentInit args)
+    {
+        component.OrbitDistance =
+            _robustRandom.NextFloat(0.75f * component.OrbitDistance, 1.25f * component.OrbitDistance);
+
+        component.OrbitLength = _robustRandom.NextFloat(0.5f * component.OrbitLength, 1.5f * component.OrbitLength);
+    }
+
+    public override void FrameUpdate(float frameTime)
+    {
+        base.FrameUpdate(frameTime);
+
+        foreach (var (orbit, sprite) in EntityManager.EntityQuery<OrbitVisualsComponent, ISpriteComponent>())
+        {
+            var angle = new Angle(Math.PI * 2 * orbit.Orbit);
+            var vec = angle.RotateVec(new Vector2(orbit.OrbitDistance, 0));
+
+            sprite.Rotation = angle;
+            sprite.Offset = vec;
+        }
+    }
+
+    protected override void OnAppearanceChange(EntityUid uid, OrbitVisualsComponent component, ref AppearanceChangeEvent args)
+    {
+        if (!args.Component.TryGetData<bool>(OrbitingVisuals.IsOrbiting, out var orbiting))
+            return;
+
+        if (!TryComp<ISpriteComponent>(uid, out var sprite))
+            return;
+
+        var animationPlayer = EntityManager.EnsureComponent<AnimationPlayerComponent>(uid);
+
+        if (orbiting)
+        {
+            if (animationPlayer.HasRunningAnimation(_orbitAnimationKey))
+                return;
+
+            animationPlayer.Play(GetOrbitAnimation(component), _orbitAnimationKey);
+        }
+        else
+        {
+            RemComp<OrbitVisualsComponent>(uid);
+            if (animationPlayer.HasRunningAnimation(_orbitAnimationKey))
+            {
+                animationPlayer.Stop(_orbitAnimationKey);
+            }
+
+            if (!animationPlayer.HasRunningAnimation(_orbitStopKey))
+            {
+                animationPlayer.Play(GetStopAnimation(component, sprite), _orbitStopKey);
+            }
+        }
```

Comment:
Should orbiting also stop the stoporbitkey?

Reply:
yeah